### PR TITLE
Enhancement range validator traversable

### PR DESCRIPTION
**Pull request by SamMousa**
Proposed implementation for #10154

### CODE REVIEW

File: framework/helpers/BaseArrayHelper.php [hunk 1]
```diff
@@ -584,4 +584,51 @@ public static function isIndexed($array, $consecutive = false)
             return true;
         }
     }
+
+    /**
+     * Checks if needle is in haystack, supports array-like objects.
+     * @see in_array
+     * @param array|\Traversable $needle
+     * @param array|\Traversable $haystack
+     * @param boolean $strict
```
Comment (SilverFire) [inline framework/helpers/BaseArrayHelper.php]: Need a description of this attribute, maybe with example

File: framework/helpers/BaseArrayHelper.php [hunk 2]
```diff
@@ -584,4 +584,51 @@ public static function isIndexed($array, $consecutive = false)
...
+        if ($haystack instanceof \Traversable) {
+            foreach($haystack as $value) {
+                if ($needle == $value && (!$strict || $needle === $haystack)) {
+                    return true;
+                }
+            }
+        } elseif (is_array($haystack)) {
+            return in_array($needle, $haystack, $strict);
+        } else {
+            throw new \InvalidArgumentException("Argument \$haystack must be an array or implement Traversable");
```
Comment (SilverFire) [inline framework/helpers/BaseArrayHelper.php]: `"` may be changed to `'` and there will be no need to escape '$'

File: framework/helpers/BaseArrayHelper.php [hunk 3]
```diff
@@ -584,4 +584,51 @@ public static function isIndexed($array, $consecutive = false)
             return true;
         }
     }
+
+    /**
+     * Checks if needle is in haystack, supports array-like objects.
```
Comment (SilverFire) [inline framework/helpers/BaseArrayHelper.php]: `needle` -> `$needle`
`haystack` -> `$haystack`

File: tests/framework/helpers/ArrayHelperTest.php [hunk 4]
```diff
@@ -478,4 +478,23 @@ public function testHtmlDecode()
             ]
         ], ArrayHelper::htmlDecode($array, false));
     }
+
+    public function testIn()
+    {
+
```
Comment (SilverFire) [inline tests/framework/helpers/ArrayHelperTest.php]: Need tests for `$strict = true`

**Comment by SamMousa** [comment]
@SilverFire, thanks for the feedback; will fix the things you mentioned after the actual implementation / approach has been approved ;-)

**Comment by SilverFire** [comment]
The code is fine as for me.
Need to add a Changelog line and update a guide page http://www.yiiframework.com/doc-2.0/guide-helper-array.html

**Comment by SamMousa** [comment]
@SilverFire  Done!

**Comment by SilverFire** [comment]
We order changelog by type (bug, enh, chg, new), then by issue ID.

Merged with changelog reword https://github.com/yiisoft/yii2/commit/22d999728174b1c29af111c53ad487717b90c2fa

**Comment by SilverFire** [comment]
Thank you!

**Comment by SamMousa** [comment]
Noted for next time.

**Pull request closed without merging**